### ISSUE TITLE
Turn on visible remote push notifications

### ISSUE THREAD

**Issue by shankari** [e-mission/e-mission-phone#210]
So that we can send out interesting notifications from the server.
The big question is that, since we already have a set of callbacks on remote push on iOS, how can we also use the notification handler from the push plugin.

In particular, if we have two places with 

`didRegisterForRemoteNotificationsWithDeviceToken` defined, will both be called, or will only one be called?

**Comment by shankari** [comment]
Added breakpoints at:

```
AppDelegate+notification.m:
- didRegisterForRemoteNotificationsWithDeviceToken
- didFailToRegisterForRemoteNotificationsWithError
- didReceiveRemoteNotification (w/o completionHandler)
- didReceiveRemoteNotification (with completionHandler)

PluginPush.m
- init (for the registration code)

BEMAppDelegate.m:
- didFinishLaunchingWithOptions (at registerForRemoteNotifications)
- didRegisterForRemoteNotificationsWithDeviceToken
- didFailToRegisterForRemoteNotificationsWithError
- didReceiveRemoteNotification (with completionHandler)
```

**Comment by shankari** [comment]
Order in which breakpoints were triggered

```
BEMAppDelegate::didFinishLaunchingWithOptions

#0	0x00000001001609bc in +[AppDelegate(notification) didFinishLaunchingWithOptions:] at /Users/shankari/e-mission/e-mission-phone/platforms/ios/emission/Plugins/edu.berkeley.eecs.emission.cordova.datacollection/BEMAppDelegate.m:52
#1	0x000000010015e3e0 in -[BEMDataCollection initWithConsent] at /Users/shankari/e-mission/e-mission-phone/platforms/ios/emission/Plugins/edu.berkeley.eecs.emission.cordova.datacollection/BEMDataCollection.m:45
#2	0x000000010015e1bc in -[BEMDataCollection pluginInitialize] at /Users/shankari/e-mission/e-mission-phone/platforms/ios/emission/Plugins/edu.berkeley.eecs.emission.cordova.datacollection/BEMDataCollection.m:31
#3	0x0000000100219588 in -[CDVViewController registerPlugin:withClassName:] at /Users/shankari/e-mission/e-mission-phone/platforms/ios/CordovaLib/Classes/Public/CDVViewController.m:562
#4	0x00000001002199bc in -[CDVViewController getCommandInstance:] at /Users/shankari/e-mission/e-mission-phone/platforms/ios/CordovaLib/Classes/Public/CDVViewController.m:602
#5	0x000000010021731c in -[CDVViewController viewDidLoad] at /Users/shankari/e-mission/e-mission-phone/platforms/ios/CordovaLib/Classes/Public/CDVViewController.m:313
#6	0x00000001000dbf34 in -[MainViewController viewDidLoad] at /Users/shankari/e-mission/e-mission-phone/platforms/ios/emission/Classes/MainViewController.m:76
```

```
AppDelegate+notification

application::didRegisterForRemoteNotification

#0	0x00000001001b677c in -[AppDelegate(notification) application:didRegisterForRemoteNotificationsWithDeviceToken:] at /Users/shankari/e-mission/e-mission-phone/platforms/ios/emission/Plugins/phonegap-plugin-push/AppDelegate+notification.m:90
#1	0x000000010059125c in _dispatch_call_block_and_release ()
#2	0x000000010059121c in _dispatch_client_callout ()
#3	0x0000000100596284 in _dispatch_main_queue_callback_4CF ()
#4	0x0000000186451f2c in __CFRUNLOOP_IS_SERVICING_THE_MAIN_DISPATCH_QUEUE__ ()
#5	0x000000018644fb18 in __CFRunLoopRun ()
#6	0x000000018637e048 in CFRunLoopRunSpecific ()
#7	0x0000000187e01198 in GSEventRunModal ()
#8	0x000000018c358628 in -[UIApplication _run] ()
#9	0x000000018c353360 in UIApplicationMain ()
#10	0x00000001000dbb40 in main at /Users/shankari/e-mission/e-mission-phone/platforms/ios/emission/main.m:32

```

```
PushPlugin.m

application::init

#0	0x00000001001baa14 in __19-[PushPlugin init:]_block_invoke at /Users/shankari/e-mission/e-mission-phone/platforms/ios/emission/Plugins/phonegap-plugin-push/PushPlugin.m:338
#1	0x000000010059125c in _dispatch_call_block_and_release ()
#2	0x000000010059121c in _dispatch_client_callout ()
#3	0x000000010059f2d4 in _dispatch_queue_override_invoke ()
#4	0x00000001005a0e6c in _dispatch_root_queue_drain ()
#5	0x00000001005a0bb8 in _dispatch_worker_thread3 ()
#6	0x00000001855352c8 in _pthread_wqthread ()
#7	0x0000000185534db4 in start_wqthread ()

```


```
AppDelegate+notification

application::didRegisterForRemoteNotification

#0	0x00000001001b677c in -[AppDelegate(notification) application:didRegisterForRemoteNotificationsWithDeviceToken:] at /Users/shankari/e-mission/e-mission-phone/platforms/ios/emission/Plugins/phonegap-plugin-push/AppDelegate+notification.m:90
#1	0x000000010059125c in _dispatch_call_block_and_release ()
#2	0x000000010059121c in _dispatch_client_callout ()
#3	0x0000000100596284 in _dispatch_main_queue_callback_4CF ()
#4	0x0000000186451f2c in __CFRUNLOOP_IS_SERVICING_THE_MAIN_DISPATCH_QUEUE__ ()
#5	0x000000018644fb18 in __CFRunLoopRun ()
#6	0x000000018637e048 in CFRunLoopRunSpecific ()
#7	0x0000000187e01198 in GSEventRunModal ()
#8	0x000000018c358628 in -[UIApplication _run] ()
#9	0x000000018c353360 in UIApplicationMain ()
#10	0x00000001000dbb40 in main at /Users/shankari/e-mission/e-mission-phone/platforms/ios/emission/main.m:32
#11	0x00000001853605b8 in start ()

```

**Comment by shankari** [comment]
Interesting to note that `didRegisterForRemoteNotification` was called twice, once for each call to `register`. I wonder if this is because the class in both cases was `AppDelegate (notification)`. Let's try changing that name...

**Comment by shankari** [comment]
Changed to `AppDelegate (DataCollection)` instead. Now, order:

```
- BEMAppDelegate::didFinishLaunchingWithOptions
- AppDelegate+notification: application::didRegisterForRemoteNotificationsWithDeviceToken
- PushPlugin::init
- AppDelegate+notification: application::didRegisterForRemoteNotificationsWithDeviceToken
```

**Comment by shankari** [comment]
Commented out `AppDelegate+notification: application::didRegisterForRemoteNotificationsWithDeviceToken`

```
- BEMAppDelegate::didFinishLaunchingWithOptions
- BEMAppDelegate: application::didRegisterForRemoteNotificationsWithDeviceToken
- PushPlugin::init
- BEMAppDelegate: application::didRegisterForRemoteNotificationsWithDeviceToken
```

**Comment by shankari** [comment]
So basically, you cannot have two - if you have both, only one will be invoked anyway.
Confirm that this is true for the `didReceiveRemoteNotification` calls by setting this to the 1min channel.

**Comment by shankari** [comment]
Just for the record, I am able to register for push notifications now but have not been able to receive any notifications on a physical device.

The device is fine - I can get notifications on a fresh install from the store. I can also get local notifications with the dev build too. But I can't get remote notifications on the dev build.

There is probably something stupid and annoying that is wrong, but I need to document. Let me document that in here so that I can reproduce in the future.

**Comment by shankari** [comment]
Now, I can receive notifications with a fresh install from the store on iphone2, but not on iphone1. I got a bunch of notifications on iphone1 earlier, but I don't get them any more. didn't work even after app update so that they are both with the same version.

Might try to uninstall and reinstall but very disappointed that this is so flaky.

**Comment by shankari** [comment]
Fresh install on iphone1 works super reliably (like overnight).
I am going to regenerate push notification certs (since I have to do that anyway) and see if that helps.

I was also able to confirm that the token was registered on ionic, and to send a notification to it via the API.

**Comment by shankari** [comment]
Hm, we currently don't have a dev certificate - the only certificate is for production. 

<img width="735" alt="screen shot 2017-02-14 at 11 15 49 am" src="https://cloud.githubusercontent.com/assets/2423263/22944870/fd284fa4-f2a6-11e6-8506-eef8e22563c7.png">

```
<plist version="1.0">
<dict>
    <key>aps-environment</key>
    <string>development</string>
</dict>
</plist>
```

I wonder if that is the reason why the notifications are not showing up on the dev server...

**Comment by shankari** [comment]
I didn't see any certificates in the developer screen (see https://github.com/e-mission/e-mission-phone/issues/210#issuecomment-279808890) but I did see a certificate in the keychain, expiring on ~ Feb 10. Unfortunately, I deleted it, but that makes it more likely that this is the problem.

**Comment by shankari** [comment]
Yup! That is the problem. I uploaded the new dev certificate to ionic, and deleted the old cert and then re-uploaded on back4app and it works too.

still dunno why release version wasn't getting updates yesterday, probably because of high load with per-minute notifications?

**Comment by shankari** [comment]
If it is a visible notification, we don't get a callback until the user clicks on it, and then we get a call to
`(void) application:(UIApplication *)application didReceiveRemoteNotification:(NSDictionary *)userInfo`
so that won't affect the interaction with the data collection, which overrides

`- (void)application:(UIApplication *)application didReceiveRemoteNotification:(NSDictionary *)userInfo fetchCompletionHandler:(void (^)(UIBackgroundFetchResult))completionHandler`

for the silent push notifications

Let's wait until 2pm when the silent push notification will trigger, and figure out which callback will be invoked.

**Comment by shankari** [comment]
At 2pm

```
- (void)application:(UIApplication *)application didReceiveRemoteNotification:(NSDictionary *)userInfo fetchCompletionHandler:(void (^)(UIBackgroundFetchResult))completionHandler {
```
was indeed called, but only on the push plugin

**Comment by shankari** [comment]
We need to think about how to structure native callbacks.

Let's first focus on iOS, since that's what we need right now.
- init() function is the equivalent of `didFinishLaunchingWithOptions`
    - takes in a config spec, parses it, registers for types etc
    - then requests token from apple
    - then gets callback in the delegate and generates event for javascript

- gets callback when remote notification is delivered
    - creates a handler block
    - set it as the handlerObj of the push handler
    - in the push handler, when finish() is called, the handler is marked complete

What we need to do:
- refactor init() into parsing config and registering for remote
- call register from our code (see `getCommandInstance` for an example).
  specify callback for registration and notification
- on register callback, call all register callbacks
- on notify callback, this is a bit trickier because we need to invoke the
  completion handler
    - currently, we pass around the completion handler (both in the push code and in the data collection code)
    - we could switch that to tasks instead. Then, when all the tasks are complete, we mark the completion handler as complete. That way, we don't have to keep track of all the task tracking and the completion handler stuff - there's a nice library that will do it for us.
    - this means that our new forked push plugin library will have a new dependency on the Bolts.framework but it already has a bunch of dependencies on various things so hopefully it is not too bad.

**Comment by shankari** [comment]
Just to clarify: we would pass around `BFTaskCompletionSource` objects since that is where we can call `setResult` to indicate that we are done. And we would add the `BFTaskCompletionSource.task` objects to an array and then call `taskForCompletionOfAllTasks` on the array.

See `BEMServerSyncCommunicationHelper` for an example.

**Comment by shankari** [comment]
Do I even have to do this from native code? in android, maybe, but we haven't felt the need to do a silent push notification in android because there is reasonable background support. But in iOS, we assume that the UI is always loaded and the javascript is always running.

So in that case, we can simply the code to listen to silent push notifications in javascript and then rebroadcast to the state machine. Why isn't that good enough?

**Comment by shankari** [comment]
It is really due to the completionHandler code. We need to know when the native code has finished processing data so that we can call the completion handler.

But why can't we just generate another notification - REMOTE_PUSH_HANDLED or something?
We can, as long as we have only one native plugin that deals with remote pushes.
If we have multiple, we would need to mark when each of them was done, and then mark the completion handler when all of them were done.

This seems that it basically matches the promise pattern - you callback all the plugins that care about silent pushes, get a promise from each and then when all the promises are done, mark them as final - i.e. something like

```
var callbackList = pluginList.map(function(plugin) {
   return plugin.handleSilentPushNotification();
}
Promise.all(callbackList).then(function(result) {
   push.finish(success);
}, function(error) {
   push.finish(error);
}
```

**Comment by shankari** [comment]
For now, thought, we have only one plugin that deals with silent push notifications. And using a callback with a promise is a bit more complex in the native code changes. And we haven't yet thought through how the android code should work.

For now, since we have only one plugin that uses this, let us not over-engineer.
We will just use the broadcast notifications for now. In that case, the change is fairly simple - the datacollection plugin just needs to generate a notification instead of calling the completion handler directly.

**Comment by shankari** [comment]
problem with doing this is that we also need to handle the background fetch case, and in that case, we do want to use the completion handler (at least as per current code).

Right now, both silent remote push and background fetch call the same 
`launchTripEndCheckAndRemoteSync` function, and that seems like a Good design pattern to follow.

So we have three possible options to do this:
1. we can expose `launchTripEndCheckAndRemoteSync` through the javascript interface. Note that if we do so, we cannot have it take the completionHandler as a parameter. Instead, it would need to do something else to indicate that the task was complete.
    1. it will not return anything, instead we will generate a local event when processing is complete. The javascript code for remote push case will listen to this directly - the native code for the background fetch case will listen to it and call the completion handler
    1. it will continue to return a completion handler, but the plugin interface will not expose it. Instead the plugin interface will pass in a completionHandler block which, when executed, will invoke the plugin completion code
1. we integrate with the background fetch handler as a plugin as well. Then, `launchTripEndCheckAndRemoteSync` moves to javascript - we invoke it (including the remote push local event) in both cases.

Option 2 seems to be the most consistent with the notion of controlling from javascript. Option 1 (ii) seems most consistent with the long-term promise-based solution. Unsure how well javascript-based control will work on android, and whether we even want to use silent push notifications on android.

Let's try 2 and see how well it works.

**Comment by shankari** [comment]
I ended up going with 1 (ii) to be more consistent with the eventual promise-based solution.
The combination of visible and invisible push notifications is tricky, so I am going to document what I have found about the push so far.

on android:
if `content_available` is set, title and message are ignored.

this generates a visible notification
```
    "notification": {
        "title": "Silent and visible",  # if title + message + content_available is set, we get both display and callback on iOS 
        "message": "shows up anyway!",  # but on android, the title and message are null!
        "android": {
            # "content_available": 1,  # note that for android, this neeeds to be with an underscore
            "data": {
               "key": "this is android raw"    
            },
            "payload": {
               "key": "this is the android payload"
            }
            # "title": "Android specific title",
            # "message": "Android specific message"
        },
```

this generates a silent notification. Note that the title and message are ignored because of the `content-available`.

```
    "notification": {
        "title": "Silent and visible",  # if title + message + content_available is set, we get both display and callback on iOS 
        "message": "shows up anyway!",  # but on android, the title and message are null!
        "android": {
            "content_available": 1,  # note that for android, this neeeds to be with an underscore
            "data": {
               "key": "this is android raw"    
            },
            "payload": {
               "key": "this is the android payload"
            }
            # "title": "Android specific title",
            # "message": "Android specific message"
        },
```

Also, if we have a general title and message and a platform specific one, the platform specific one wins (as expected). In this example, the title was "Android specific title" and "Android specific message".

```
    "notification": {
        "title": "Silent and visible",  # if title + message + content_available is set, we get both display and callback on iOS 
        "message": "shows up anyway!",  # but on android, the title and message are null!
        "android": {
            # "content_available": 1,  # note that for android, this neeeds to be with an underscore
            "data": {
               "key": "this is android raw"    
            },
            "payload": {
               "key": "this is the android payload"
            },
            "title": "Android specific title",
            "message": "Android specific message"
        },
```

**Comment by shankari** [comment]
On iOS, including the title and the message and content handles both. The title and message are visually displayed even before the silent push notification callback is invoked.

```
    "notification": {
        "title": "Silent and visible",  # if title + message + content_available is set, we get both display and callback on iOS 
        "message": "shows up anyway!",  # but on android, the title and message are null!
        "ios": {
            "content_available": 1,  # note that for iOS also, this needs to be with an underscore. The code converts it to dash
            "priority": 10,
            "data": {
                "key": "this is ios raw"
            },
            "payload": {
               "key": "this is the android payload"
            }
        }
```

**Comment by shankari** [comment]
Something weird is happening with the silent push notifications on iOS.
For the first silent push, everything works as expected:
- we get the silent push callback in javascript
- we invoke handleSilentPush
- we get the promise response in javascript
- we call finish

But on the second silent push, something weird happens:
- we get the silent push callback in javascript
- we invoked handleSilentPush
- handleSilentPush completes, but we don't get the promise response in javascript
- so we never call finish

| Working | Not working |
| --------- | --------- |
| ![screen shot 2017-02-19 at 6 22 30 pm](https://cloud.githubusercontent.com/assets/2423263/23109742/800ab3e4-f6d0-11e6-90a4-2fea9f03526d.png) | ![screen shot 2017-02-19 at 6 22 38 pm](https://cloud.githubusercontent.com/assets/2423263/23109741/8007ffd2-f6d0-11e6-94b2-387c7a912799.png) |

**Comment by shankari** [comment]
https://github.com/e-mission/e-mission-phone/issues/210#issuecomment-280972790 is because `handleRemotePush` is called twice with the same callbackId. Note that the callback ID (`DataCollection1106212265`) is the same for both invocations.

| Working | Not working |
| -------- | ------------ |
| ![screen shot 2017-02-19 at 7 24 47 pm](https://cloud.githubusercontent.com/assets/2423263/23110843/304b7204-f6d9-11e6-8753-a1168bdbdd39.png) | ![screen shot 2017-02-19 at 7 24 54 pm](https://cloud.githubusercontent.com/assets/2423263/23110842/3048e5ca-f6d9-11e6-97f8-b4120b6adc84.png) |

**Comment by shankari** [comment]
Hm, this is weird. I added the following log statement to cordova.js

```
    // Register the callbacks and add the callbackId to the positional
    // arguments if given.
    if (successCallback || failCallback) {
        callbackId = service + cordova.callbackId++;
       console.log("After incrementing callbackId, id is "+callbackId);
        cordova.callbacks[callbackId] =
            {success:successCallback, fail:failCallback};
    }
```

But, as we saw, both invocations are to the same callbackId.

```
2017-02-19 19:01:59.576096 emission[1448:781429] BEMDataCollection:pluginInitialize singleton -> initialize statemachine and delegate
2017-02-19 19:02:47.583906 emission[1448:781429] After incrementing callbackId, id is DataCollection1106212265
2017-02-19 19:02:54.400811 emission[1448:781429] in handleSilentPush, sending result to id DataCollection1106212265
2017-02-19 19:02:54.402216 emission[1448:781429] INFO: in handleSilentPush, sending result to id DataCollection1106212265
2017-02-19 19:02:54.428925 emission[1448:781429] in handleSilentPush, sending result to id DataCollection1106212265
2017-02-19 19:02:54.430747 emission[1448:781429] DEBUG: in handleSilentPush, sending result to id DataCollection1106212265
2017-02-19 19:03:16.146602 emission[1448:781429] After incrementing callbackId, id is DataCollection1106212271
2017-02-19 19:03:25.167716 emission[1448:781429] in handleSilentPush, sending result to id DataCollection1106212265
2017-02-19 19:03:25.169579 emission[1448:781429] INFO: in handleSilentPush, sending result to id DataCollection1106212265
2017-02-19 19:03:25.192970 emission[1448:781429] in handleSilentPush, sending result to id DataCollection1106212265
2017-02-19 19:03:25.195827 emission[1448:781429] DEBUG: in handleSilentPush, sending result to id DataCollection1106212265
```

[reuse_callbackid.logs.gz](https://github.com/e-mission/e-mission-phone/files/786490/reuse_callbackid.logs.gz)

**Comment by shankari** [comment]
https://github.com/e-mission/e-mission-phone/issues/210#issuecomment-280980001
is because the block appears to cache the callbackId from the first invocation.

First invocation:

```
2017-02-19 19:58:36.416747 emission[1468:790754] Invoking method with callbackId = DataCollection641875063
2017-02-19 19:58:36.417365 emission[1468:790754] handleSilentPush outside the try block = DataCollection641875063
---
2017-02-19 19:58:41.062845 emission[1468:790754] in handleSilentPush, sending result to id DataCollection641875063
2017-02-19 19:58:41.063157 emission[1468:790754] INFO: in handleSilentPush, sending result to id DataCollection641875063
2017-02-19 19:58:41.090805 emission[1468:790754] in handleSilentPush, sending result to id DataCollection641875063
2017-02-19 19:58:41.091243 emission[1468:790754] DEBUG: in handleSilentPush, sending result to id DataCollection641875063
```

Second invocation:

```
2017-02-19 19:58:58.405438 emission[1468:790754] Invoking method with callbackId = DataCollection641875069
2017-02-19 19:58:58.405760 emission[1468:790754] handleSilentPush outside the try block = DataCollection641875069
---
2017-02-19 19:59:05.332174 emission[1468:790754] in handleSilentPush, sending result to id DataCollection641875063
2017-02-19 19:59:05.333151 emission[1468:790754] INFO: in handleSilentPush, sending result to id DataCollection641875063
2017-02-19 19:59:05.358505 emission[1468:790754] in handleSilentPush, sending result to id DataCollection641875063
2017-02-19 19:59:05.359638 emission[1468:790754] DEBUG: in handleSilentPush, sending result to id DataCollection641875063
```

**Comment by shankari** [comment]
Aha! This is because we set the handler if it is not yet set, but we never set it to nil.

```
    @synchronized([BEMRemotePushNotificationHandler instance]) {
        if (self.silentPushHandler == NULL && [note.object isEqualToString:CFCTransitionRecievedSilentPush]) {
            // We are the silent push handler, so we store the handler block, but don't do anything else.
            NSDictionary* userInfo = note.userInfo;
            self.silentPushHandler = [userInfo objectForKey:@"handler"];
            [LocalNotificationManager addNotification:[NSString stringWithFormat:@"Setting the SILENT_PUSH handler block in the silent push handler"]];
        }
    }
    if (self.silentPushHandler != NULL) {
        [LocalNotificationManager addNotification:[NSString stringWithFormat:
                                                   @"Received notification %@ while processing silent push notification", note.object] showUI:FALSE];
...
        // TODO: Figure out whether we should set it to NULL here or whether parts of
        // the system will still try to access the handler.
        // _silentPushHandler = nil;
    } else {
```

I am not sure how this ever worked. Wouldn't we just have set the wrong completion handler?

Here are the relevant log statements

```
2017-02-19 20:10:37.362816 emission[1475:793336] Setting the SILENT_PUSH handler block in the silent push handler
2017-02-19 20:10:37.362936 emission[1475:793336] DEBUG: Setting the SILENT_PUSH handler block in the silent push handler
2017-02-19 20:10:37.379708 emission[1475:793336] Received notification T_RECEIVED_SILENT_PUSH while processing silent push notification
2017-02-19 20:10:37.379811 emission[1475:793336] DEBUG: Received notification T_RECEIVED_SILENT_PUSH while processing silent push notification
2017-02-19 20:10:37.399826 emission[1475:793336] Handler block has already been set, ignoring SILENT_PUSH in the silent push handler
2017-02-19 20:10:37.399927 emission[1475:793336] DEBUG: Handler block has already been set, ignoring SILENT_PUSH in the silent push handler
....
2017-02-19 20:10:39.336514 emission[1475:793336] Received notification T_DATA_PUSHED while processing silent push notification
2017-02-19 20:10:39.337003 emission[1475:793336] DEBUG: Received notification T_DATA_PUSHED while processing silent push notification
2017-02-19 20:10:44.087404 emission[1475:793336] in handleSilentPush, sending result to id DataCollection1494027974
2017-02-19 20:10:44.087672 emission[1475:793336] INFO: in handleSilentPush, sending result to id DataCollection1494027974
...
2017-02-19 20:11:08.037801 emission[1475:793336] Received notification T_RECEIVED_SILENT_PUSH while processing silent push notification
2017-02-19 20:11:08.038434 emission[1475:793336] DEBUG: Received notification T_RECEIVED_SILENT_PUSH while processing silent push notification
2017-02-19 20:11:08.055059 emission[1475:793336] Handler block has already been set, ignoring SILENT_PUSH in the silent push handler
2017-02-19 20:11:08.055132 emission[1475:793336] DEBUG: Handler block has already been set, ignoring SILENT_PUSH in the silent push handler
...
2017-02-19 20:11:07.147529 emission[1475:793336] Before invoking launchTripEndCheckAndRemoteSync, id = DataCollection1494027980
2017-02-19 20:11:17.280303 emission[1475:793336] in handleSilentPush, sending result to id DataCollection1494027974
2017-02-19 20:11:17.281551 emission[1475:793336] INFO: in handleSilentPush, sending result to id DataCollection1494027974
```

**Comment by shankari** [comment]
Changed it to store a list and clear the list once the notification is done.

| Working | Working |
| -------- | -------- |
| ![screen shot 2017-02-19 at 9 34 05 pm](https://cloud.githubusercontent.com/assets/2423263/23113200/5ac31584-f6eb-11e6-95fc-274da5cdc3c1.png) | ![screen shot 2017-02-19 at 9 34 11 pm](https://cloud.githubusercontent.com/assets/2423263/23113199/5ac1bc2a-f6eb-11e6-8184-6029563f3afa.png) |

**Comment by shankari** [comment]
The silent push notifications are now 100% reliable in iOS iff the app is currently being debugged on the phone.

I have now had it go through the entire workflow at 4 fairly long intervals.

```
2017-02-20 22:12:51.285961 emission[1681:917796] In TripDiaryStateMachine, received transition T_RECEIVED_SILENT_PUSH in state STATE_WAITING_FOR_TRIP_START
2017-02-20 22:42:07.365757 emission[1681:917796] In TripDiaryStateMachine, received transition T_RECEIVED_SILENT_PUSH in state STATE_WAITING_FOR_TRIP_START
2017-02-20 23:05:00.487518 emission[1681:917796] In TripDiaryStateMachine, received transition T_RECEIVED_SILENT_PUSH in state STATE_WAITING_FOR_TRIP_START
2017-02-20 23:49:51.538098 emission[1681:917796] In TripDiaryStateMachine, received transition T_RECEIVED_SILENT_PUSH in state STATE_WAITING_FOR_TRIP_START
```

I am still not sure if the loop through javascript works if it is disconnected from the laptop, though.
So I am disconnecting the phone (and turning off debugging) now at around midnight.
I will send a silent push tomorrow morning and see what I get...

**Comment by shankari** [comment]
just confirmed that https://github.com/e-mission/e-mission-phone/issues/210#issuecomment-281270162 is true even when app is in the background - e.g. not visible on screen.

**Comment by shankari** [comment]
worked! will retry periodically through the day to see if it works consistently

**Comment by shankari** [comment]
Documented working!

| All working    |   Yay!            | 
| ------------- | ------------ |
| ![screen shot 2017-02-21 at 6 59 47 am](https://cloud.githubusercontent.com/assets/2423263/23188235/d5083632-f842-11e6-889a-03e14612c12c.png) | ![screen shot 2017-02-21 at 8 49 39 am](https://cloud.githubusercontent.com/assets/2423263/23188236/d50aea44-f842-11e6-82c2-768bc9a32722.png) |
| ![screen shot 2017-02-21 at 12 00 08 pm](https://cloud.githubusercontent.com/assets/2423263/23188238/d5111d60-f842-11e6-8a84-1def9ff57a2f.png) | ![screen shot 2017-02-21 at 2 33 21 pm](https://cloud.githubusercontent.com/assets/2423263/23188237/d50ee1f8-f842-11e6-9ab2-996f87cc73f6.png) |

**Comment by shankari** [comment]
One final fix:
- we cannot register directly on consent any more. If the app is being installed for the first time, then we don't have the login information, so we don't have the token ready to update the token. We also need to add an intro_done (or at least, logged in) check.

In particular, when I was testing the install, I switched from the install pointing to the local server to the install pointing to the production server. But the old creds were still in the keychain, so when we tried to update, we used the old credentials, which didn't work. And so the update didn't happen and I got this error while trying to send a survey.

```
INFO:root:After parsing, uuid list = [UUID('c76a0487-7e5a-3b17-a449-47be666b36f6')]
DEBUG:root:Getting tokens matching query {'user_id': {'$in': [UUID('c76a0487-7e5a-3b17-a449-47be666b36f6')]}}
Traceback (most recent call last):
  File "bin/push/send_survey.py", line 39, in <module>
    dev = args.dev)
  File "/mnt/e-mission/e-mission-server/emission/net/ext_service/push/notify_usage.py", line 13, in send_visible_notification_to_users
    token_list = pnq.get_matching_tokens(pnq.get_user_query(user_id_list))
  File "/mnt/e-mission/e-mission-server/emission/net/ext_service/push/notify_queries.py", line 27, in get_matching_tokens
    ret_list = map(lambda e: e["device_token"], ret_cursor)
  File "/mnt/e-mission/e-mission-server/emission/net/ext_service/push/notify_queries.py", line 27, in <lambda>
    ret_list = map(lambda e: e["device_token"], ret_cursor)
KeyError: 'device_token'
```

**Comment by shankari** [comment]
Going to add the intro done check and event instead...

**Comment by shankari** [comment]
re-opening this briefly to record issues with ongoing use of push notifications.
So I tried send a push notification to the test phone yesterday to get a screenshot for the slides, and it didn't work. I then tried:
- sending the push notification to dev, since I had installed from xcode
- uninstalling and re-installing from the store

Neither worked. I then tried push notifications throughout the day, and they didn't work.
Finally, today morning, I tried sending a notification while generating device logs, and it looks like it wasn't even received/processed!!

```
e-mission-phone shankari$ grep APNS /tmp/remote_push_not_working_production_install | wc -l
       0
e-mission-phone shankari$ grep emission /tmp/remote_push_not_working_production_install | wc -l
       0
```

I am going to reinstall and see if I can get a valid working version to compare.

[remote_push_not_working_production_install.gz](https://github.com/e-mission/e-mission-phone/files/831914/remote_push_not_working_production_install.gz)
